File: 11-circular-linked-list/01-traverse-circular-linked-list/01-traverse-circular-linked-list.ts
import { ListNode } from '../../TypeScriptDataStructure/CircularLinkedList/ListNode'

const traverse_circular_linked_list = (head: ListNode) => {
  if (head === null)
    return []
  let temp = head
  const res = [temp.val]
  temp = temp.next
  while (temp !== head) {
    res.push(temp.val)
    temp = temp.next
  }
  return res
}

const head = new ListNode(10, new ListNode(20, new ListNode(30, new ListNode(40))))
head.next.next.next.next = head
console.log(traverse_circular_linked_list(head))

File: 11-circular-linked-list/01-traverse-circular-linked-list/01-traverse-circular-linked-list.rb
require_relative '../../RubyDataStructure/CircularLinkedList/ListNode.rb'

def traverse_circular_linked_list(head)
  return [] if !head
  temp = head
  res = [temp.val]
  temp = temp.next
  while temp != head
    res.append(temp.val)
    temp = temp.next
  end
  return res
end

head = ListNode.new(1, ListNode.new(2, ListNode.new(3, ListNode.new(4))))
head.next.next.next.next = head

print traverse_circular_linked_list head